Add PrivatePIDs= (cont)

ryantimwilson opened the pull request:
This is a continuation of @DaanDeMeyer's PR https://github.com/systemd/systemd/pull/34174 to add a new option `PrivatePIDs=`. Unfortunately I can't merge my pull request into the existing pull request, so I created a new one.

Changes I've made on top of Daan's changes:
1. Changes dbus PrivatePIDs API to string
2. Add integration tests
3. Support for using PrivatePIDs= in unprivileged user sessions
4. Move PID namespace before mounts to ensure we mount procfs after
5. Disallow Type=forking services from using PrivatePIDs=yes
6. Ensure PrivatePIDs=yes implies MountAPIVFS=yes to always mount /proc
7. Add documentation

Note I added this feature into v258 as I expect this to take several rounds of review and I am still adding integration tests to test the feature matrix.

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,44 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+static int setup_private_pids(const ExecContext *c, ExecParameters *p) {
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        int r, q;
+
```
we typically validate function parameters with assert() for programming mistakes, i.e. add here:

```
assert(c);
assert(p);
```

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,44 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+        }
+
+        p->pidref_transport_fd = safe_close(p->pidref_transport_fd);
+
```
Can you add a big fat comment here, that highlights the fact that this function returns twice. that is an absolute exception in our codebase.

i.e. add here:

```c
/* NOTE! This function returns TWICE, once in the parent and once in the child, and it's the *child* that is going to survive, while the parent will exit shortly! */
```

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,44 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+                if (p->pidref_transport_fd >= 0) {
+                        q = send_one_fd_iov(
+                                        p->pidref_transport_fd,
+                                        pidref.fd,
```
note that at this point pidfd support is not guaranteed to be available in our tree, this should handle gracefully the case when we have no pidfd to send. (and in that case just send the pid_t, and accept that things are racy)

poettering replied:
hmmm, so something is wrong here. the receving side in pid1 doesn#t actually expect any fds attached to the message it instead looks at SO_PASSPIDFD instead afaics. but that means it must be confused by which process and is parent and which is the child?

anyway, this seems qutie broken, the sender and the receiver seem to implement different protocols.

poettering replied:
note that SO_PASSPIDFD is a very recent kernel feature. I think it would be OK to rely on its existance for this systemd feature, given this is a new feature too. But we need to make sure that we fail cleanly on old systems here. And that means that if `pidref.fd` is unset we return an error early, instead of sending an invalid fd.

ryantimwilson replied:
> note that at this point pidfd support is not guaranteed to be available in our tree, this should handle gracefully the case when we have no pidfd to send. (and in that case just send the pid_t, and accept that things are racy)

I think per your point below, we should probably just require PIDFD support to guarantee it won't be racy. In Twine agent, we hit a bunch of racy issues monitoring containers due to PID reuse and its painful.

> hmmm, so something is wrong here. the receving side in pid1 doesn#t actually expect any fds attached to the message it instead looks at SO_PASSPIDFD instead afaics. 

Okay I looked through the manager code and it looks like we use SO_PASSPIDFD to get the PID via ucreds but we also expect a PIDFD to be sent in the message body.

My guess is Daan did this to allow for dealing with kernels without SO_PASSPIDFD. The confusing part is we compare to make sure the ucreds and PIDFD sent via message body match...so it implies we still need SO_PASSPIDFD?

Anyways, yea let me dig more into this, thanks for calling this out.

```
        CMSG_FOREACH(cmsg, &msghdr) {
                if (cmsg->cmsg_level != SOL_SOCKET)
                        continue;

                if (cmsg->cmsg_type == SCM_CREDENTIALS && cmsg->cmsg_len == CMSG_LEN(sizeof(struct ucred))) {
                        assert(!ucred);
                        ucred = CMSG_TYPED_DATA(cmsg, struct ucred);
                } else if (cmsg->cmsg_type == SCM_PIDFD) {
                        assert(pidfd < 0);
                        pidfd = *CMSG_TYPED_DATA(cmsg, int);
                }
        }
```

> note that SO_PASSPIDFD is a very recent kernel feature. I think it would be OK to rely on its existance for this systemd feature, given this is a new feature too. But we need to make sure that we fail cleanly on old systems here. And that means that if pidref.fd is unset we return an error early, instead of sending an invalid fd.

I kinda like this better as opposed to relying on sending a PIDFD twice if SO_PASSPIDFD is enabled. If I go this approach, I can add comments in systemd.exec

ryantimwilson replied:
Okay ended up fixing this - you were right, we were missing parsing SCM_RIGHTS message - and added a comment in manager.c about the protocol

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4890,6 +4931,27 @@ int exec_invoke(
+                        r = setup_private_pids(context, params);
+                        if (ERRNO_IS_NEG_PRIVILEGE(r))
+                                log_exec_warning_errno(context, params, r,
+                                                       "PrivatePIDs=yes is configured, but pid namespace setup failed, ignoring: %m");
```
maybe mention "due to lack of privileges" or so in the log msg.

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4890,6 +4931,27 @@ int exec_invoke(
+
+                        if (r > 0)
+                                /* If we're in the parent process, return immediately in order to exit ASAP.
+                                 * The rest of the setup will continue in the child process. */
```
maybe make the comment clearer: start with saying "Return value > 0 means we are in the parent process"

poettering commented on src/core/dbus-execute.c:
```diff
@@ -1186,6 +1186,7 @@ const sd_bus_vtable bus_exec_vtable[] = {
         SD_BUS_PROPERTY("PrivateUsersEx", "s", property_get_private_users_ex, offsetof(ExecContext, private_users), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("PrivateMounts", "b", bus_property_get_tristate, offsetof(ExecContext, private_mounts), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("PrivateIPC", "b", bus_property_get_bool, offsetof(ExecContext, private_ipc), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("PrivatePIDs", "b", bus_property_get_bool, offsetof(ExecContext, private_pids), SD_BUS_VTABLE_PROPERTY_CONST),
```
I think we should make this an enum right-away, even if we only have on and off initially. But given the history with the other options such as PrivateUsers and ProtectControlgroups I think it would make a ton of sense to start out with a string property and a proper enum right-away, so that we don't have to do this "Ex" property business later.

poettering commented on src/core/manager.c:
```diff
@@ -1309,6 +1311,53 @@ static int manager_setup_handoff_timestamp_fd(Manager *m) {
+
+                r = setsockopt_int(m->pidref_transport_fds[0], SOL_SOCKET, SO_PASSCRED, true);
+                if (r < 0)
+                        return log_error_errno(r, "SO_PASSCRED failed: %m");
```
maybe say on which socket in the error message

poettering commented on src/core/manager.c:
```diff
@@ -1309,6 +1311,53 @@ static int manager_setup_handoff_timestamp_fd(Manager *m) {
+                        return log_error_errno(r, "SO_PASSCRED failed: %m");
+
+                r = setsockopt_int(m->pidref_transport_fds[0], SOL_SOCKET, SO_PASSPIDFD, true);
+                if (r < 0 && r != -ENOPROTOOPT)
```
maybe better ERRNO_IS_NEG_NOT_SUPPORTED()

poettering commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -3263,7 +3263,7 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
       readonly b PrivateIPC = ...;
       @org.freedesktop.DBus.Property.EmitsChangedSignal("const")
-      readonly b PrivatePIDs = ...;
+      readonly s PrivatePIDs = '...';
```
oha, can you just squash this fixup patch into the first commit. just mention in the commit msg that Daan did the original work.

poettering commented:
uh, this confused me while reviewing, because i rviewed the first patch no knowing the 2nd patch was a fixup patch for this. please squash the two (and retain a comment regarding Daan's contributions in the commit msg).

ryantimwilson commented:
> uh, this confused me while reviewing, because i rviewed the first patch no knowing the 2nd patch was a fixup patch for this. please squash the two (and retain a comment regarding Daan's contributions in the commit msg).

@poettering yea sorry about that. I was trying to ensure @DaanDeMeyer gets proper credit but ended up introducing more confusion. Will squash and fix!

bluca commented:
> > uh, this confused me while reviewing, because i rviewed the first patch no knowing the 2nd patch was a fixup patch for this. please squash the two (and retain a comment regarding Daan's contributions in the commit msg).
> 
> @poettering yea sorry about that. I was trying to ensure @DaanDeMeyer gets proper credit but ended up introducing more confusion. Will squash and fix!

Add `Co-authored-by: Daan De Meyer <daan.j.demeyer@gmail.com>` to the commit message and Github will automatically parse it and attribute credit

ryantimwilson commented:
Hmmm it seems the unprivileged tests fail in CI while passing in Centos 10 VM via mkosi locally...will debug more

poettering commented:
lgtm

ryantimwilson commented:
@poettering thanks!

Unfortunately,  unprivileged tests fail in CI while passing in Centos 10 VM via mkosi locally. So since I can't repro locally, I'm adding a bunch of debug logging here to diagnose the issue.

I'll prefix the PR title with [CI FAILING] for now to let people know not to review. I'll re-request review from you once I fix the unpriv changes.

YHNdnzj commented:
Hmm? Is this targeting v258?

ryantimwilson commented:
@YHNdnzj the plan was to target 258 since I figured this would take more rounds of review. But...since its already almost there, we can probably hit 257? I guess depends on how long for me to fix the unprivileged issue. What do you think?

YHNdnzj commented on src/core/manager.c:
```diff
@@ -1309,6 +1311,53 @@ static int manager_setup_handoff_timestamp_fd(Manager *m) {
+                        return log_error_errno(r, "Failed to enable SO_PASSCRED for pidref socket: %m");
+
+                r = setsockopt_int(m->pidref_transport_fds[0], SOL_SOCKET, SO_PASSPIDFD, true);
+                if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
```
The logging seems messed up here. We should log at warning if NOT EOPNOTSUPP, no?

YHNdnzj replied:
And I'd still debug log if pidfd is not available.

YHNdnzj commented on src/core/manager.c:
```diff
@@ -4999,6 +5056,140 @@ static int manager_dispatch_handoff_timestamp_fd(sd_event_source *source, int fd
+                r = pidref_set_pid(&parent_pidref, ucred->pid);
+        if (r < 0) {
+                if (r == -ESRCH)
+                        log_debug_errno(r, "Pidref child process died before message is processed. Ignoring.");
```
Please use proper casing, PidRef/pidref

YHNdnzj commented on src/core/manager.c:
```diff
@@ -4999,6 +5056,140 @@ static int manager_dispatch_handoff_timestamp_fd(sd_event_source *source, int fd
+                return 0;
+        }
+
+        /* Need to handle kernels without PIDFD support. */
```
dunno. If we target v258 all this compat logic can go

ryantimwilson replied:
I think we can target 257 if I can fix the unprivileged test issue soon-ish

ryantimwilson replied:
Changing PR to target 257

yuwata commented:
Not sure if it is possible, but how about the following ?
- unshare mount namespace,
- PID1 sends a mountfd of clean /proc to the child process, and the child mount it on its namespace,
- unshare pid namespace,
- unshare mount namespace again and setup mount namespace as usual.

ryantimwilson commented:
> Not sure if it is possible, but how about the following ?

@yuwata the problem with this is we are in an unprivileged user manager, so this isn't PID1. This means to mount anything, we need to join a new user namespace and once we're in a new user namespace, we can't mount a restricted /proc. To be clear, everything works fine in privileged mode in all cases. In unprivileged mode, it only doesn't work if /proc is masked.

One possible option is the user manager to call PID1 dbus to get a clean /proc mountfd...but that fails somewhat like a privilege escalation security issue. After all, there is a reason the kernel has this restriction.

Note nspawn also has this problem, just add `mount -t tmpfs tmpfs /proc/scsi` and see TEST-13-NSPAWN fail on unpriv tests. The only mystery is why are CI tests passing for nspawn...my theory is they are skipped. Pushing an update and running to CI to confirm that now. Once I confirm that, I plan to fix tests and go with option 1 above, documenting this like crazy.

ryantimwilson commented:
Okay so it appears the reason TEST-13-NSPAWN unpriv tests pass on some architectures is they specify TEST_NO_NSPAWN=1, meaning they do not run in an nspawn container but qemu.

And guess who is masking /proc/kmsg and /proc/sys/kernel/random/boot_id...nspawn lol.

Anyways, let me add a big ol comment and fix the tests. But this was an interesting investigation

ryantimwilson commented:
@poettering would appreciate your re-review

Updates:
1. Documented issue with /proc masking for unprivileged user namespaces (see above for details) and fixed integration tests
2. Changed to version 257
3. Address @YHNdnzj comments

poettering commented on man/systemd.exec.xml:
```diff
@@ -1963,6 +1963,29 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
+        <citerefentry><refentrytitle>systemd-nspawn</refentrytitle><manvolnum>1</manvolnum></citerefentry> does),
+        a new <filename>/proc</filename> will not be mounted in the new PID namespace and processes from the parent
+        PID namespace will still be visible in <filename>/proc</filename>. This is due to a kernel restriction not
+        allowing unprivileged user namespaces to mount a less restrictive instance of <filename>/proc</filename>.</para>
```
hmm, i think we should not allow this: run a service in a pidns with the host procfs mounted it, the distinctions are just too many. 

I think we should in the long run reconsider in nspawn whether we want to overmount /proc/ like this (or at least make it an option).

I think we should treat an overmounted /proc/ the same way as pidns not available due to perms really.

ryantimwilson replied:
@poettering yea for user experience, I agree.

The problem I was having is we unshare the PID namespace prior to the mount namespace and mounting /proc. So we don't know whether the /proc mount will fail until we're already in the PID namespace.

I have a few possible solutions:
1. Unshare the mount namespace, set up all mounts, then only if /proc mount succeeded, unshare the PID namespace + remount /proc. IMHO, this will be challenging because after calling apply_mount_namespace(), we're already in the container root filesystem and we want to avoid mutating it (e.g. create temporary mount point, mount new proc and its child mounts there, move them over).
2. Before unsharing the PID namespace, try to detect if /proc is over-mounted. This is complicated because mounts over empty dirs are allowed (e.g. /proc/sys/fs/binfmt_misc) while other mounts over non-empty dirs/files are not (e.g. /proc/kmsg). I feel this would be brittle
3. Before unsharing the PID namespace, clone a temporary short-lived process into a new mount namespace, enable slave propagation, then try to mount /proc. If /proc mount fails, we know to not unshare the PID namespace. We have some precedent for doing this as we fork off a temporary process to set up user namespaces for private users.

I vote 3 but happy to hear other opinions/options.

poettering replied:
yeah, option 3 sounds good to me. kinda sucks, but it is how it is.

ryantimwilson replied:
@poettering fyi fun fact, we have to unshare both the mount namespace AND PID namespace in the temporary subprocess. It seems the kernel doesn't let unprivileged user namespaces mount a new proc unless they're in a new PID namespace, otherwise we get more EACCES errors

Anyways, got it to work, pushing soon, sorry for adding more reviews here :/

poettering commented on src/core/manager.c:
```diff
@@ -4999,6 +5058,140 @@ static int manager_dispatch_handoff_timestamp_fd(sd_event_source *source, int fd
+        if (msghdr.msg_flags & MSG_TRUNC) {
+                log_warning("Got truncated pidref message, ignoring.");
+                return 0;
+        }
```
this check is not necessary, recvmsg_safe() these days turns MSG_TRUNC into -EXFULL

poettering commented on src/core/manager.c:
```diff
@@ -4999,6 +5058,140 @@ static int manager_dispatch_handoff_timestamp_fd(sd_event_source *source, int fd
+        if (ERRNO_IS_NEG_TRANSIENT(n))
+                return 0; /* Spurious wakeup, try again */
+        if (n == -EXFULL) {
+                log_warning("Got message with truncated control, ignoring.");
```
truncated control is ECHRNG, not EXFULL.

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,128 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+        /* Create a communication channel so that the child can tell the parent a proper error code in case it
+         * failed. */
+        if (pipe2(errno_pipe, O_CLOEXEC) < 0)
+                return -errno;
```
maybe also log_exec_debug_errno()?

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,128 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+         * only applies to user managers, we skip this check for other scopes.
+         */
+        if (p->runtime_scope != RUNTIME_SCOPE_USER)
+                return 1;
```
that's not strictly true. in theory we do support systems without CAP_SYS_ADMIN (i.e. in a container world, where the container manager sets up *all* mounts, so that we don't have to anymore).

maybe do return have_effective_capability(CAP_SYS_ADMIN)?

ryantimwilson replied:
Odd this isn't working as expected. `have_effective_cap(CAP_SYS_ADMIN)` is returning true for unprivileged user managers but fails to mount /proc. Hence tests are failing. Let me debug further

ryantimwilson replied:
Okay another fun fact (these are getting less fun as time goes on): the kernel checks that the process has CAP_SYS_ADMIN within the user namespace corresponding to the mount namespace. So we actually need to do this check in the temporary process.

ryantimwilson replied:
Fixed by removing this check altogether

ryantimwilson replied:
Changed this to check CAP_SYS_ADMIN prior to entering userns to avoid PID1 re-checking if it can mount /proc over and over. Seems like a waste

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,128 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+        if (n != 0) /* on success we should have read 0 bytes */
+                return -EIO;
+
+        r = wait_for_terminate_and_check("(sd-proc-check)", TAKE_PID(pid), 0);
```
prefix last arg with `/* flags= */` as per our current coding style

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,128 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+                /* If something strange happened with the child, let's consider this fatal, too */
+                log_exec_debug(c, p, "Child process (sd-proc-check) exited with unexpected exit status '%d'.", r);
+                return -EIO;
+        }
```
do:

```c
if (r != EXIT_SUCCESS)
        /* if something … */
        return log_exec_debug(c, p, SYNTHETIC_ERRNO(EIO), "Child proc…");
```

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4873,6 +5001,32 @@ int exec_invoke(
+                                * The rest of the setup will continue in the child process. */
+                                return 0;
+                } else if (r == 0)
+                        log_exec_warning(context, params, "PrivatePIDs=yes is configured, but /proc cannot be re-mounted due to lack of privileges, ignoring.");
```
in log msgs and docs we usually want that path names that refer to dirs are suffixed with a slash. i.e. /proc → /proc/

ryantimwilson replied:
will fix docs too

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4873,6 +5001,32 @@ int exec_invoke(
+                } else if (r == 0)
+                        log_exec_warning(context, params, "PrivatePIDs=yes is configured, but /proc cannot be re-mounted due to lack of privileges, ignoring.");
+                else
+                        log_exec_warning_errno(context, params, r, "PrivatePIDs=yes is configured, but failed to detect if /proc can be remounted, ignoring: %m");
```
ditto

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4873,6 +5001,32 @@ int exec_invoke(
+                        log_exec_warning(context, params, "PrivatePIDs=yes is configured, but /proc cannot be re-mounted due to lack of privileges, ignoring.");
+                else
+                        log_exec_warning_errno(context, params, r, "PrivatePIDs=yes is configured, but failed to detect if /proc can be remounted, ignoring: %m");
+        }
```
so I think that failure to get PrivatePIDs=yes should fail the service. Most security related options we handle gracefully, under the assumption that not applying them just gives you more rights but won't change execution in a meaningful way. But I think for PrivatePIDs= that#s not the case. It very much changes execution, because PID 1 has quite different semantics from other PIDs (they are target for reparenting and such), moreover things like writing pid files (for those legacy daemons that do that) becomes really wweird, because what they'd write there might not match what the rest of the system sees.

if you want this to be graceful, a middle ground might be to add PrivatePIDs=try or so, so that people explicitly opt into graceful handling of this knob. But if you don#t need this I wouldn't bother and just make inability to apply privatepids a failure here.

Compare how BindPath= failing will casue a service failure for us, but ReadOnlyPath= is handled gracefully. That also follows the logic that the former rearrranges things too heavily to ignore if it doesnt work. The latter only takes away privs.

ryantimwilson replied:
Yea at least for me, we don't need/want graceful fallback. So I will remove it.

ryantimwilson commented:
Last issue appears to be sanitizer tests are broken: https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/18532/

Working on this now, will ping when fixed

ryantimwilson commented:
@poettering @DaanDeMeyer turns out systemd-executor (parent process) hangs at exit() on __lsan::DoLeakCheck. Unfortunately, I'm having trouble setting `detect_leaks=0` in LSAN_OPTIONS as in tests, its set via kernel command line and reloading/reexecing systemd with DefaultEnvironment set doesn't seem to work.

So unless anyone objects, I'm gonna do what we do in some other integration tests and not run these under ASAN for now, then follow up later.

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,126 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+                 *
+                 * Unlike services, mount points, sockets and swap devices do not use pidref
+                 * transport. */
+                if (p->pidref_transport_fd >= 0) {
```
Check this earlier and fail otherwise? This is our internal IPC and should always be around

ryantimwilson replied:
It will not always be around unfortunately because pidref_transport_fd  is only set for services, not for other exec types (mount/socket/swap device)

ryantimwilson replied:
See:
```
        if (UNIT_VTABLE(u)->notify_pidref)
                p->pidref_transport_fd = u->manager->pidref_transport_fds[1];
```

in unit.c

YHNdnzj replied:
Please just refuse PrivatePIDs= in the corresponding `_verify` funcs in that case.

ryantimwilson replied:
Ack then I will add the verification that p->pidref_transport_fd >= 0

YHNdnzj commented on src/basic/process-util.h:
```diff
@@ -193,6 +193,7 @@ typedef enum ForkFlags {
         FORK_DETACH             = 1 << 18, /* Double fork if needed to ensure PID1/subreaper is parent */
         FORK_NEW_NETNS          = 1 << 19, /* Run child in its own network namespace                             💣 DO NOT USE IN THREADED PROGRAMS! 💣 */
         FORK_PACK_FDS           = 1 << 20, /* Rearrange the passed FDs to be FD 3,4,5,etc. Updates the array in place (combine with FORK_CLOSE_ALL_FDS!) */
+        FORK_NEW_PIDNS          = 1 << 21, /* Run child in its own PID namespace                                 💣 DO NOT USE IN THREADED PROGRAMS! 💣 */
```
Maybe group ns flags together while at it?

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,126 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+         * where unprivileged user namespaces cannot mount a less restrictive instance of /proc. */
+
+        /* Create temporary directory for mounting /proc. */
+        r = mkdtemp_malloc("/tmp/systemd-temporary-proc-XXXXXX", &tmpdir);
```
Is this dir really needed? Can probably use some known existing dir, e.g. /dev/shm/? (Our credential logic uses that in the intermediate mount ns too)

ryantimwilson replied:
/dev/shm works for me!

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,126 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+        if (r != EXIT_SUCCESS) {
+                /* If something strange happened with the child, let's consider this fatal, too */
+                return log_exec_debug_errno(c, p, SYNTHETIC_ERRNO(EIO), "Child process (sd-proc-check) exited with unexpected exit status '%d'.", r);
+        }
```
Unneeded brackets

YHNdnzj commented on src/core/manager.c:
```diff
@@ -5002,6 +5061,140 @@ static int manager_dispatch_handoff_timestamp_fd(sd_event_source *source, int fd
+                return 0;
+        }
+
+        if (!UNIT_VTABLE(u)->notify_pidref)
```
Additionally this should warn about unexpected messages

YHNdnzj commented on src/core/service.c:
```diff
@@ -710,6 +710,9 @@ static int service_verify(Service *s) {
                 return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(ENOEXEC), "Service is of type D-Bus but no D-Bus service name has been specified. Refusing.");
 
+        if (s->type == SERVICE_FORKING && exec_needs_pid_namespace(&s->exec_context))
+                return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(ENOEXEC), "Service of type Forking does not support PrivatePIDs=yes. Refusing.");
```
Why uppercase?

ryantimwilson replied:
The rest of the logs in service_verify are like that e.g.
```
Service has Restart= set to either always or on-success, which isn't allowed for Type=oneshot services. Refusing.
```

YHNdnzj replied:
(sorry for the confusion, I was referring to "forking")

ryantimwilson replied:
oh yes this should be `Type=forking`, will fix

YHNdnzj commented on src/core/service.c:
```diff
@@ -4908,6 +4911,24 @@ static void service_handoff_timestamp(
+
+        if (pidref_equal(&s->main_pid, parent_pidref)) {
+                pidref_done(&s->main_pid);
+                s->main_pid = TAKE_PIDREF(*child_pidref);
```
service_set_main_pidref?

ryantimwilson replied:
@YHNdnzj I think the problem is service_set_main_pidref makes an assumption the main PID will be our child, otherwise it logs and sets main_pid_alien: https://github.com/systemd/systemd/blob/89099136d7c28310d7159817c90e970c62a995ec/src/core/service.c#L245

In PID namespaces, main PID won't be our child but its still running systemd-executor. So I don't think we want to set main_pid_alien

The rest of the function looks useful e.g. adjusting starting timestamp. Maybe I will refactor to use part of this function.

ryantimwilson replied:
Actually...since its not our child, I think we do want to set main_pid_alien = true since we won't get SIGCHLD if it exits. So lets call this function as is

YHNdnzj commented on src/core/service.c:
```diff
@@ -4908,6 +4911,24 @@ static void service_handoff_timestamp(
+                s->main_pid = TAKE_PIDREF(*child_pidref);
+        } else if (pidref_equal(&s->control_pid, parent_pidref)) {
+                pidref_done(&s->control_pid);
+                s->control_pid = TAKE_PIDREF(*child_pidref);
```
This is not watched through unit_watch_pidref?

ryantimwilson replied:
Yea seems we should, this is missing

bluca commented:
> @poettering @DaanDeMeyer turns out systemd-executor (parent process) hangs at exit() on __lsan::DoLeakCheck, which is not surprising given the well-known issues with LSAN/ASAN and forking. Unfortunately, I'm having trouble setting `detect_leaks=0` in LSAN_OPTIONS as in tests, its set via kernel command line and reloading/reexecing systemd with DefaultEnvironment set doesn't seem to work.
> 
> So unless anyone objects, since we want to cut 257 soon, I'm gonna do what we do in some other integration tests and not run these under ASAN for now, then follow up later.
> 
> However, if you feel strongly we need to fix this prior, happy to hold off.

But it works for other cases though, right? So it must be due to some interaction with the pid namespace? If you can't figure it out now it's fine, but please do file an issue to track it, otherwise it will be forgot

ryantimwilson commented:
@bluca I filed an issue and I will look at it next week: https://github.com/systemd/systemd/issues/34992

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2252,120 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+         * where unprivileged user namespaces cannot mount a less restrictive instance of /proc. */
+
+        /* Create temporary directory for mounting /proc. */
+        r = mkdtemp_malloc("/dev/shm/systemd-temporary-proc-XXXXXX", &tmpdir);
```
Hmm, what I was suggesting was to directly mount over /dev/shm/. The dir is used in new mountns only, hence shouldn't have any impact on /dev/shm/ on host. This saves us from any sort of explicit tmpdir creation/cleanup.

ryantimwilson replied:
Oh sorry, I didn't get that. Yea makes sense, I'll change

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4877,6 +5001,44 @@ int exec_invoke(
+                                *exit_status = EXIT_NAMESPACE;
+                                return log_exec_error_errno(context, params, r, "Failed to set up pid namespace: %m");
+                        }
+
```
can you drop this empty line btw, we try to keep function calls and their error/return handling together. hence place empty lines before func calls, and after their error handling/return handling but not in the middle like this.

poettering commented:
So hmm, I think we should just call _exit() in the parent, if we keep only the child around. and not do the "clean" return path in that case.

I mean, usually it's the other way round: the parent is the one that stays around and cleans things up, but here we reverse the roles, hence it should only be the child that cleans things up if necessary.

I think that would both be more correct and fix the asan issue too.

Various of our destructors have external effects, hence you want to call them only once. In this current version of the patch you call them twice however, which is quite problematic I think

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,121 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+                /* This is the expected case where proc cannot be mounted due to permissions. */
+                if (ERRNO_IS_NEG_PRIVILEGE(r))
+                        return 0;
+                else if (r < 0)
```
drop the `else`, the word is redundant

poettering commented on src/core/exec-invoke.c:
```diff
@@ -4877,6 +5001,44 @@ int exec_invoke(
+                        if (r > 0)
+                                /* Return value > 0 means we are in the parent process. Thus, return immediately in order to exit ASAP.
+                                * The rest of the setup will continue in the child process. */
+                                return 0;
```
so, I think this should actually be an `_exit(EXIT_SUCCESS)`, as discussed above.

ryantimwilson replied:
Done! Had to add some logic to send errno from parent -> child via pipe2. Its similar to can_mount_proc() but we can't have the child wait for the parent so we always have to send the errno.

Confirmed this fixes the ASAN issue, thanks!

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -4924,6 +5065,39 @@ int exec_invoke(
+                        *exit_status = EXIT_NAMESPACE;
+                        return log_exec_error_errno(context, params, SYNTHETIC_ERRNO(EPERM),
+                                                    "PrivatePIDs=yes is configured, but /proc/ cannot be re-mounted due to lack of privileges, refusing.");
+                } else {
```
I'd always check error (< 0) first, then the special case (== 0), that way the normal path forward can be moved out of if branch

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,138 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+        if (pipe2(errno_pipe, O_CLOEXEC) < 0)
+                return -errno;
+
+        r = pidref_safe_fork("(sd-pidfd-child)", FORK_NEW_PIDNS, &pidref);
```
pidfd -> pidns?

YHNdnzj commented on src/core/service.c:
```diff
@@ -4908,6 +4911,38 @@ static void service_handoff_timestamp(
+        Service *s = ASSERT_PTR(SERVICE(u));
+        int r;
+
+        assert(parent_pidref);
```
pidref_is_set?

YHNdnzj commented on src/core/swap.c:
```diff
@@ -256,6 +256,9 @@ static int swap_verify(Swap *s) {
         if (!unit_has_name(UNIT(s), e))
                 return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(ENOEXEC), "Value of What= and unit name do not match, not loading.");
 
+        if (exec_needs_pid_namespace(&s->exec_context))
```
Hmm, actually this can be further dedupped by checking unit_vtable->notify_pidref in unit_verify_contexts I think. But feel free to ignore.

YHNdnzj commented on src/core/service.c:
```diff
@@ -4908,6 +4911,38 @@ static void service_handoff_timestamp(
+        if (pidref_equal(&s->main_pid, parent_pidref)) {
+                r = service_set_main_pidref(s, TAKE_PIDREF(*child_pidref), /* start_timestamp = */ NULL);
+                if (r < 0)
+                        return log_unit_warning_errno(u, r, "Failed to set new main pid: %m");
```
Hmm, if this fails, the error would be propagated to manager_dispatch_..., and error in such handlers will result in the event source getting disabled. Please just make this a void, and always return 0 in handler.

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,138 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+        /* Try to read an error code from the child */
+        n = read(errno_pipe[0], &r, sizeof(r));
+        if (n < 0)
+                return log_exec_debug_errno(c, p, -errno, "Failed to read errno from pipe with child process (sd-proc-check): %m");
```
drop the `-` from the `errno`. It's redundant. the logging functions all do `abs()` on the passed errors anyway

poettering commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,138 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+         * receive an errno even on success. */
+        n = read(errno_pipe[0], &r, sizeof(r));
+        if (n < 0)
+                return log_exec_debug_errno(c, p, -errno, "Failed to read errno from pipe with parent process: %m");
```
as above: `-errno` → `errno`

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2251,6 +2251,137 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
+         * where unprivileged user namespaces cannot mount a less restrictive instance of /proc. */
+
+        /* Verify /dev/shm/ exists. */
+        r = is_dir("/dev/shm/", /* follow = */ true);
```
The check is kinda pointless too. Let's always assume this exists (and this shall never be a symlink, so at least follow should be disabled)

YHNdnzj commented on src/core/manager.c:
```diff
@@ -1309,6 +1311,55 @@ static int manager_setup_handoff_timestamp_fd(Manager *m) {
+
+                r = setsockopt_int(m->pidref_transport_fds[0], SOL_SOCKET, SO_PASSPIDFD, true);
+                if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                        log_debug_errno(r, "SO_PASSPIDFD is not supported for pidref socket, ignoring: %m");
```
If the message already indicates errno clearly, there's no need to log with %m

ryantimwilson replied:
Changed it to log_debug and removed %m

YHNdnzj commented on src/core/manager.c:
```diff
@@ -5002,6 +5061,142 @@ static int manager_dispatch_handoff_timestamp_fd(sd_event_source *source, int fd
+        }
+
+        if (!UNIT_VTABLE(u)->notify_pidref) {
+                log_unit_debug(u, "Received pidref event from unexpected unit type '%s'.", unit_type_to_string(u->type));
```
This should be at warning level?

YHNdnzj commented on src/core/service.c:
```diff
@@ -4908,6 +4911,39 @@ static void service_handoff_timestamp(
+                r = service_set_main_pidref(s, TAKE_PIDREF(*child_pidref), /* start_timestamp = */ NULL);
+                if (r < 0) {
+                        log_unit_warning_errno(u, r, "Failed to set new main pid: %m");
+                        return;
```
You can simplify this with return (void) log_... BTW

YHNdnzj commented on src/core/service.c:
```diff
@@ -4908,6 +4911,39 @@ static void service_handoff_timestamp(
+                if (r < 0)
+                        log_unit_warning_errno(u, r, "Failed to watch new control PID " PID_FMT ": %m", s->control_pid.pid);
+        } else {
+                log_unit_debug(u, "Parent process " PID_FMT " does not match main or control processes", parent_pidref->pid);
```
```suggestion
                log_unit_debug(u, "Parent process " PID_FMT " does not match main or control processes, ignoring.", parent_pidref->pid);
```

YHNdnzj reviewed (commented):
LGTM, just minor stuff now really

Winterhuman commented:
Just to double check, does this pull include an entry for `PrivatePIDs=` under `systemd-analyze security`, or is that not included here? Should I open an issue for it if not?

bluca commented:
it's not included, please do file an issue for it
